components/shared/net/lib.rs

```diff
@@ -945,5 +945,31 @@ pub fn http_percent_encode(bytes: &[u8]) -> String {
     percent_encoding::percent_encode(bytes, HTTP_VALUE).to_string()
 }
 
+pub fn user_info_percent_encode<T: AsRef<[u8]>>(bytes: T) -> String {
```

Comment:
@jdm I followed [the spec](https://url.spec.whatwg.org/#userinfo-percent-encode-set) but still a bit unsure if I did this right. 

Right now I'm not sure if the usage of `CONTROLS` conflicts with its usage in the function `http_percent_encode` found right above this function. I'm unsure because `CONTROLS` is a constant and I don't know if it retains characters added to it after its been used and whether that affects future uses.

Reply:
I believe rust-url already implements this (which feels like the right layer of abstraction to me): https://github.com/servo/rust-url/blob/7492360d4230b67fa0e62794b6fde276525e5f84/url/src/lib.rs#L2301 . We should be able to call `set_username` on the url without performing any preprocessing on the username we pass in.

Reply:
oh nice, perfect